doc/rtd/explanation/format.rst

```diff
@@ -24,93 +59,132 @@ These things may include:
+may run once per instance, every boot, or once ever. See the associated
+module to determine the run frequency.
 
-Begins with: ``#cloud-config`` or ``Content-Type: text/cloud-config`` when
-using a MIME archive.
-
-.. note::
-   Cloud config data can also render cloud instance metadata variables using
-   :ref:`jinja templates <instancedata-Using>`.
+See the :ref:`yaml_examples` section for a set of commented examples of
+supported cloud config formats.
 
 .. _user_data_script:
 
 User data script
 ================
 
-Typically used by those who just want to execute a shell script.
+Example
+-------
+
+.. code-block:: shell
+
+    #!/bin/sh
+    echo "Hello World" > /var/tmp/output.txt
 
-Begins with: ``#!`` or ``Content-Type: text/x-shellscript`` when using a MIME
-archive.
+Explanation
+-----------
 
-User data scripts can optionally render cloud instance metadata variables using
-:ref:`jinja templates <instancedata-Using>`.
+A user data script is a single shell script to be executed once per instance.
+User data scripts are run relatively late in the boot process, during the
+cloud-init's final stage as part of the "cc_scripts_user" module. When run,
+the environment variable ``INSTANCE_ID`` is set to the current instance ID
+for use within the script.
 
-Example script
+.. _user_data_formats-cloud_boothook:
+
+Cloud boothook
+==============
+
+Simple Example
 --------------
 
-Create a script file :file:`myscript.sh` that contains the following:
+.. code-block:: shell
 
-.. code-block::
+   #cloud-boothook
+   #!/bin/sh
+   echo 192.168.1.130 us.archive.ubuntu.com > /etc/hosts
 
+Example of once-per-instance script
+-----------------------------------
+
+.. code-block:: bash
+
+   #cloud-boothook
    #!/bin/sh
-   echo "Hello World.  The time is now $(date -R)!" | tee /root/output.txt
 
-Now run:
+   PERSIST_ID=/var/lib/cloud/first-instance-id
+   _id=""
+   if [ -r $PERSIST_ID ]; then
+     _id=$(cat /var/lib/cloud/first-instance-id)
+   fi
 
-.. code-block:: shell-session
+   if [ -z $_id ]  || [ $INSTANCE_ID != $_id ]; then
+     echo 192.168.1.130 us.archive.ubuntu.com >> /etc/hosts
+   fi
+   sudo echo $INSTANCE_ID > $PERSIST_ID
 
-   $ euca-run-instances --key mykey --user-data-file myscript.sh ami-a07d95c9
+Explanation
+-----------
 
-Kernel command line
-===================
+A cloud boothook is similar to a :ref:`user data script<user_data_script>`
+in that it is a shell script run on boot. When run,
+the environment variable ``INSTANCE_ID`` is set to the current instance ID
+for use within the script.
 
-When using the NoCloud datasource, users can pass user data via the kernel
-command line parameters. See the :ref:`NoCloud datasource<datasource_nocloud>`
-and :ref:`explanation/kernel-command-line:Kernel command line` documentation
-for more details.
+The boothook is different in that:
 
-Gzip compressed content
-=======================
+* It is run very early in boot, during the 'init' stage, before any
```

Comment:
Could we link to the init stage on the boot stages page from here? That would be useful for anybody who isn't familiar with cloud-init's ordering.